examples/raft-kv-rocksdb/src/network/raft_network_impl.rs

```diff
@@ -72,7 +72,7 @@ impl RaftNetworkFactory<ExampleTypeConfig> for ExampleNetwork {
     ) -> Result<Self::Network, Self::ConnectionError> {
         dbg!(&node);
         let addr = format!("ws://{}", node.rpc_addr);
-        let client = Client::dial_websocket(&addr).await.ok();
+        let client = Some(Client::dial_websocket(&addr).await.map_err(|e| NetworkError::new(&e))?);
         Ok(ExampleNetworkConnection { addr, client, target })
```

Comment:
Hmm.. `connect()` should return something that can `dial` when needed, but not an established connection.

In this example app, `connect()` will be called only once, when the replication stream is created. And a replication stream won't be created again unless the leader or the membership changes.

This means, there is no chance for the leader to **reconnect** to a follower if the follower went down and went up again, after the replication stream is created.

The `connect()` should return an error only when there is no chance a **reconnect** would succeed. E.g., the configured address for this follower is invalid.

🤔 It looks like `connect()` is a little misleading. What about `new_client()` or something similar?

Reply:
If the function isn't supposed to connect then perhaps new_client would make more sense.

This makes it a lot more complicated too, as it opens up the question of how to deal with the "bad" client if we know it can never be valid. Perhaps splitting it out in a two-step process would make sense? Having a `validate_node` callback to the application that refuses to add nodes that fails it and rename `connect` to `new_client` and make it unfailable again?

Reply:
> This makes it a lot more complicated too, as it opens up the question of how to deal with the "bad" client if we know it can never be valid. 

Right. It's kind of application-specific.

> Perhaps splitting it out in a two-step process would make sense? Having a `validate_node` callback to the application that refuses to add nodes that fails it and rename `connect` to `new_client` and make it unfailable again?

I'd prefer not to add a new API unless it has to.

If an application needs to validate a node, it can validate the `Node` just when [`add_learner()`](https://github.com/datafuselabs/openraft/blob/70e3318abb13aca60b91f98dfcedba2bcb99ee0e/openraft/src/core/raft_core.rs#L460) is called.

Leaving the return value of `connect()` a `Result` is still useful to me:)
An application may have trouble building a new client.

Reply:
that's sensdible, how should we deal with `new_client` erroring in that case, should ikt shut down the raft core as it is a unrecoverable error?

Reply:
RaftCore will just ignore the replication stream that it can not `connect()` to.
And RaftCore still works with a minority of replication streams down.

- There is a test about this behavior in the PR: #488

Reply:
Fair point, I just checked an invalid `Config` also won't cause RaftCore to stop, that makes the behavior of allowing invalid nodes consistant.

so should I change the PR to just rename `connect` to `new_client` and add some big warning note in the docs about errors?

Reply:
> Fair point, I just checked an invalid `Config` also won't cause RaftCore to stop, that makes the behavior of allowing invalid nodes consistant.
> 
> so should I change the PR to just rename `connect` to `new_client` and add some big warning note in the docs about errors?

That would be great:DDD